src/relayclient/AccountManager.ts

```diff
@@ -43,34 +54,52 @@ export default class AccountManager {
   // TODO: make forwarder part of RelayRequest, why is it dangling??
   async sign (relayRequest: RelayRequest, forwarderAddress: Address): Promise<PrefixedHexString> {
     let signature
-    const signedData = await getDataToSign({
+    const signedData = getDataToSign({
       chainId: this.chainId,
       verifier: forwarderAddress,
       relayRequest
     })
     const keypair = this.accounts.find(account => isSameAddress(account.address, relayRequest.relayData.senderAddress))
     if (keypair != null) {
-      // @ts-ignore
-      signature = sigUtil.signTypedData_v4(this.ephemeralKeypair.privateKey, { data: signedData })
+      signature = this._signWithControlledKey(keypair, signedData)
     } else {
-      signature = await getEip712Signature(
-        {
-          web3: this.web3,
-          methodSuffix: this.config.methodSuffix ?? '',
-          jsonStringifyRequest: this.config.jsonStringifyRequest ?? false,
-          dataToSign: signedData
-        })
+      signature = await this._signWithExternallyControlledKey(signedData)
+    }
+
+    // Sanity check
+    let rec: Address
+    try {
+      // @ts-ignore
+      rec = sigUtil.recoverTypedSignature_v4({
+        data: signedData,
+        sig: signature
+      })
+    } catch (error) {
+      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+      throw new Error(`Failed to sign relayed transaction for ${relayRequest.relayData.senderAddress}`)
     }
-    // Sanity check only
-    // @ts-ignore
-    const rec = sigUtil.recoverTypedSignature_v4({
-      data: signedData,
-      sig: signature
-    })
     if (!isSameAddress(relayRequest.relayData.senderAddress.toLowerCase(), rec)) {
       // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
-      throw new Error(`signature error: sender=${relayRequest.relayData.senderAddress}, recovered=${rec}`)
+      throw new Error(`Internal RelayClient exception: signature is not correct: sender=${relayRequest.relayData.senderAddress}, recovered=${rec}`)
     }
     return signature
   }
+
+  // These methods is extracted to
+  // a) allow different implementations in the future, and
+  // b) allow spying on Account Manager in tests
+  async _signWithExternallyControlledKey (signedData: any): Promise<string> {
```

Comment:
I'm not sure about the method naming: its not "external", but using provider's sign method (that is, provider-controlled accounts)

Reply:
Give me a name then.

Reply:
signWithProviderAccount

Reply:
blocker: "_signWithExternallyControlledKey" is bad name, especially since it hides where the signature actually comes from.
(the provider call is buried inside getEip712Signature)